src/Dialect/ONNX/DisposablePool.hpp

```diff
@@ -0,0 +1,81 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===-------------------------- DisposablePool.hpp ------------------------===//
+//
+// DisposablePool manages instances of DisposableElementsAttr.
+// It creates them, maintains a record of them (in a "pool") until they are
+// deemed unreachable, and it can be called to garbage collect those that are
+// unreachable, and to "scrub" all occurrences in a module by replacing each
+// with a DenseElementsAttr.
+//
+// Garbage collected and scrubbed DisposableElementsAttrs are removed from the
+// pool and their reference to the underlying MemoryBuffer is cleared,
+// decrementing the shared_ptr reference count.
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "src/Dialect/ONNX/DisposableElementsAttr.hpp"
+
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/DialectInterface.h"
+
+#include <unordered_map>
+
+namespace onnx_mlir {
+
+class DisposablePool : public mlir::DialectInterface::Base<DisposablePool> {
+  friend class ElementsAttrBuilder; // allow access to insert()
+public:
+  static DisposablePool &create(mlir::MLIRContext *context);
+
+  static DisposablePool *get(mlir::MLIRContext *context);
+
+  // Disposes every DisposableElementsAttr and in moduleOp replaces each with a
+  // DenseElementsAttr.
+  static void scrub(mlir::ModuleOp moduleOp, DisposablePool *disposablepool);
+
+  DisposablePool(mlir::Dialect *dialect, mlir::MLIRContext *context);
+  ~DisposablePool();
+
+  mlir::DisposableElementsAttr lookup(size_t id) const;
+
+  // Disposes every DisposableElementsAttr in the pool which is unreachable
+  // (doesn't appear in moduleOp).
+  void garbageCollectUnreachable(mlir::ModuleOp moduleOp);
+
+  void close() {
+    assert(pool.empty() && "pool must be scrubbed before close");
```

Comment:
Should there be a `close` in the destructor, to make sure it is not simply lost? Or maybe there is no need as there is one obj for the duration of the compilation unit.

Reply:
no need as there is one obj for the duration of the compilation unit

and I don't think it's necessary to insist that all instances are scrubbed or garbage collected, I think there could be valid use cases for that